convex/game/agents.ts

```diff
@@ -385,78 +351,53 @@ export const agentFindConversation = internalMutation({
     if (!player) {
       throw new Error(`Couldn't find player: ${args.playerId}`);
     }
-    const location = await ctx.db.get(player.locationId);
-    if (!location) {
-      throw new Error(`Couldn't find location: ${player.locationId}`);
+    const agent = await ctx.db.get(args.agentId);
+    if (!agent) {
+      throw new Error(`Couldn't find agent: ${args.agentId}`);
     }
-    const position = { x: location.x, y: location.y };
+    // Don't try to start a new conversation if we were just in one.
+    const justLeftConversation =
+      agent.lastConversation && now < agent.lastConversation + CONVERSATION_COOLDOWN;
 
-    const otherPlayers = await ctx.db
-      .query('players')
-      .withIndex('active', (q) => q.eq('worldId', args.worldId).eq('active', true))
-      .filter((q) => q.neq(q.field('_id'), args.playerId))
-      .collect();
+    // Don't try again if we recently tried to find someone to invite.
+    const recentlyAttempted =
+      agent.lastInviteAttempt && now < agent.lastInviteAttempt + CONVERSATION_COOLDOWN;
 
-    const candidates = [];
+    const recentActivity = player.activity && now < player.activity.until + ACTIVITY_COOLDOWN;
 
-    for (const otherPlayer of otherPlayers) {
-      // Skip players that are currently in a conversation.
-      const member = await conversationMember(ctx.db, otherPlayer._id);
-      if (member) {
-        continue;
-      }
-      // Find the latest conversation we're both members of.
-      const lastMember = await ctx.db
-        .query('conversationMembers')
-        .withIndex('left', (q) =>
-          q
-            .eq('playerId', args.playerId)
-            .eq('status.kind', 'left')
-            .eq('status.with', otherPlayer._id),
-        )
-        .order('desc')
-        .first();
-
-      if (lastMember) {
-        if (lastMember.status.kind !== 'left') {
-          throw new Error(`Unexpected status: ${lastMember.status.kind}`);
-        }
-        if (now < lastMember.status.ended + PLAYER_CONVERSATION_COOLDOWN) {
-          continue;
-        }
-      }
-      const location = await ctx.db.get(otherPlayer.locationId);
-      if (!location) {
-        throw new Error(`Couldn't find location: ${otherPlayer.locationId}`);
+    // Decide whether to do an activity or
+    if (!player.pathfinding) {
+      if (recentActivity || justLeftConversation) {
+        // TODO: decide where to go
+        // await insertInput(ctx, args.worldId, 'finishDoSomething', {
+        //   uuid: args.uuid,
+        //   agentId: args.agentId,
+        //   destination:
+        // });
+      } else {
+        // TODO: have LLM choose the activity & emoji
```

Comment:
one question structurally is whether we should change this upfront to be an action that calls a query rather than a mutation. 

it'll probably help with OCC load too since then we don't have to worry about the engine "preempting" the decision mutation with its periodic writes.

Reply:
Totally - and if it passed in the player & agent documents, instead of just the IDs, we could make it this far without needing an upfront query

Reply:
Actually without bringing in a helper to make the document validator for the player / agent, it's easier to stick it in a query for now